DataStructures/Heap/FibonacciHeap.cs

```diff
@@ -0,0 +1,595 @@
+        protected void CascadingCut(FHeapNode y)
+        {
+            var z = y.Parent;
+            if (z != null)
+            {
+                if (y.Mark == false)
+                {
+                    y.Mark = true;
+                }
+                else
+                {
+                    Cut(y, z);
+                    CascadingCut(z);
+                }
+            }
+        }
+
+        /// <summary>
+        /// <para>
+        /// Consolidate is analogous to Heapify in <see cref="DataStructures.Heap.BinaryHeap{T}"/>.
+        /// </para>
+        /// <para>
+        /// First, an array <c>A</c> [0...D(H.n)] is created where H.n is the number
+        /// of items in this heap, and D(x) is the max degree any node can have in a
+        /// Fibonacci heap with x nodes.
+        /// </para>
+        /// <para>
+        /// For each node <c>x</c> in the root list, try to add it to <c>A[d]</c> where
+        /// d is the degree of <c>x</c>.
+        /// If there is already a node in <c>A[d]</c>, call it <c>y</c>, and make
+        /// <c>y</c> a child of <c>x</c>. (Swap <c>x</c> and <c>y</c> beforehand if
+        /// <c>x</c> is greater than <c>y</c>). Now that <c>x</c> has one more child,
+        /// remove if from <c>A[d]</c> and add it to <c>A[d+1]</c> to reflect that its
+        /// degree is one more. Loop this behavior until we find an empty spot to put
+        /// <c>x</c>.
+        /// </para>
+        /// <para>
+        /// With <c>A</c> all filled, empty the root list of the heap. And add each item
+        /// from <c>A</c> into the root list, one by one, making sure to keep track of
+        /// which is smallest.
+        /// </para>
+        /// </summary>
+        protected void Consolidate()
+        {
+            if (MinItem == null)
+            {
+                return;
+            }
+
+            // There's a fact in Intro to Algorithms:
+            // "the max degree of any node in an n-node fibonacci heap is O(lg(n)).
+            // In particular, we shall show that D(n) <= floor(log_phi(n)) where phi is
+            // the golden ratio, defined in equation 3.24 as phi = (1 + sqrt(5))/2"
+            //
+            // For a proof, see [1]
+            var maxDegree = 1 + (int)Math.Log(Count, (1 + Math.Sqrt(5)) / 2);
+
+            // Create slots for every possible node degree of x
+            var a = new FHeapNode?[maxDegree];
+            var siblings = SiblingIterator(MinItem).ToList();
+            foreach (var w in siblings)
+            {
+                var x = w;
+                var d = x.Degree;
+
+                // While A[d] is not empty, we can't blindly put x here
+                while (a[d] != null)
+                {
+                    var y = a[d];
+
+                    // This is just here to satisfy C#, otherwise it complains that y
+                    // might be null further below.
```

Comment:
Does it really complain? It seems obvious that a[d] != null & y = a[d] => y != null. It's strange if the compiler can't figure this out, but ok... Please double-check, just to make sure.

Reply:
This is what I'm seeing, it could be the case that C# is unwilling to assume because it can't be sure if the value at `A[i]` has changed in between the first access and here. I think I'll assign to `y` outside of the loop and do a null check on that, it seems like that eliminates the need for a second null check.

![image](https://user-images.githubusercontent.com/10136926/109730949-5cf10480-7b88-11eb-8267-1b2404aeb48a.png)

Reply:
I checked it locally and it seems that the nullability state of array elements is not tracked, only of stand-alone variables. It's fine to have this check here, or another possibility is to write x.Key.CompareTo(y!.Key) > 0 (note the **!**)